### PR TITLE
Fix ring of experts using TP

### PR DESCRIPTION
**Pull request by NuojCheng**
# Description

When `use_ring_of_experts=True` and TP is enabled, a sharding shape error appears. This PR fixes it by adding constraint on the unpermute shapes.

FIXES: b/490177598

# Tests

CI tests

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/moe.py [hunk 1]
```diff
@@ -1318,7 +1318,7 @@ def get_active_sharding_axes(pspec_dim_axes, tensor_dim_index):
         )
 
         # Sum up the partial outputs across the expert shards.
-        output = jnp.reshape(output, (-1, sequence_length, self.config.emb_dim))
+        output = jnp.reshape(output, (-1, sequence_length, self.config.emb_dim // self.get_tensor_parallelism_size()))
```
Comment (suexu1025) [inline src/maxtext/layers/moe.py:1321]: Do we need manually rs gradient output?

Comment (NuojCheng) [reply src/maxtext/layers/moe.py:1321]: If we rely on JAX autodiff then no. If custom vjp then probably yes.
